discovery/manager.go

```diff
@@ -296,6 +296,49 @@ func (m *Manager) startProvider(ctx context.Context, p *Provider) {
 
 // cleaner cleans resources associated with provider.
 func (m *Manager) cleaner(p *Provider) {
+
+	// Before deleting the targets of p from m.targets definitely, we should ask
+	// the consumers to forget about them
+
+	// this logic should go somewhere else as this will block the reload
+	// maybe add a m.droppedTargets and make sender() expire them?
+	// maybe make allGroups to that? as it keeps empty targets
+
+	tSets := map[string][]*targetgroup.Group{}
+	m.targetsMtx.Lock()
+	p.mu.RLock()
+	for pkey, tsets := range m.targets {
+		// TODO: m.subs can at most contain one sub, don't see wht the plural and why a map? (maybe easy to use?)
+		for s := range p.subs {
+			key := poolKey{s, p.name}
+			for _, tg := range tsets {
+				// The targets were discovered by the provider
+				if pkey == key {
+					// Expire them
+					// works without the Source?? What is Source used for?
+					// or we can just drop them.
+					tSets[s] = append(tSets[s], &targetgroup.Group{Source: tg.Source})
+				} else {
+					// Keep the ones discovered by others
+					tSets[s] = append(tSets[s], tg)
+				}
+			}
+		}
+	}
+	p.mu.RUnlock()
+	m.targetsMtx.Unlock()
+
+	// TODO: updates metrics, etc.
+	select {
+	case m.syncCh <- tSets:
+		fmt.Println("Cleaner set targets to drop for provider", p.name)
+	default:
+		// TODO: Retry when needed
```

Comment:
especially when it makes sense (channel can get closed...)

Reply:
what can we do then? is there a valid case?

Reply:
retry in case the channel is full but not when when the scrape manager is stopped e.g.

I will think about this, and as I mentioned in comments, I'm not sure we want to keep this logic here inside the ApplyConfig...

Reply:
Went for the other solution I proposed: do that in `allGroups`, check out the last changes.

Reply:
I added other unit tests as well.